packages/astro/src/core/build/static-build.ts

```diff
@@ -367,19 +362,27 @@ async function runPostBuildHooks(
 /**
  * For each statically prerendered page, replace their SSR file with a noop.
  * This allows us to run the SSR build only once, but still remove dependencies for statically rendered routes.
+ * If a component is shared between a statically rendered route and a SSR route, it will still be included in the SSR build.
  */
 async function cleanStaticOutput(
 	opts: StaticBuildOptions,
 	internals: BuildInternals,
 	ssrOutputChunkNames: string[]
 ) {
 	const allStaticFiles = new Set();
-	for (const pageData of eachPageData(internals)) {
-		if (pageData.route.prerender && !pageData.hasSharedModules) {
-			const { moduleSpecifier } = pageData;
-			const pageBundleId = internals.pageToBundleMap.get(moduleSpecifier);
-			const entryBundleId = internals.entrySpecifierToBundleMap.get(moduleSpecifier);
+	const allSSRFiles = new Set();
+	for (const pageData of internals.pagesByKeys.values()) {
+		const { moduleSpecifier } = pageData;
+		const pageBundleId = internals.pageToBundleMap.get(moduleSpecifier);
+		const entryBundleId = internals.entrySpecifierToBundleMap.get(moduleSpecifier);
+		if (pageData.route.prerender && !pageData.hasSharedModules && !allSSRFiles.has(pageBundleId ?? entryBundleId)) {
 			allStaticFiles.add(pageBundleId ?? entryBundleId);
+		} else {
+			allSSRFiles.add(pageBundleId ?? entryBundleId);
+			// Check if the component was not previously added to the static build by a statically rendered route
+			if (allStaticFiles.has(pageBundleId ?? entryBundleId)) {
```

Comment:
We are computing **four times** `pageBundleId ?? entryBundleId`. Let's do it once and save it in a variable.